text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1827 @@
+      [rfcs#1993][rfc_opaque_data_structs].
+    - `DynSized` as implemented is similar to that from
+      [rfcs#1993][rfc_opaque_data_structs] except it is implemented for every
+      type with a known size and alignment at runtime, rather than requiring an
+      `Opaque` type.
+    - In addition to preventing extern types being used in `size_of_val` and
+      `align_of_val`, this PR is motivated by wanting to have a mechanism by which
+      `!DynSized` types can be prevented from being valid in struct tails due to needing
+       to know the alignment of the tail in order to calculate its field offset.
+    - `DynSized` had to be made a implicit supertrait of all traits in this
+      implementation - it is presumed this is necessary to avoid unsized types
+      implementing traits.
+    - This actually went through FCP and would have been merged if not
+      eventually closed for inactivity.
+- [rust#46108: Add DynSized trait (rebase of #44469)][pr_dynsized_rebase], [mikeyhew][author_mikeyhew], Nov 2017
+    - This pull request is a resurrection of [rust#44469][pr_dynsized].
+    - Concerns were raised about the complexity of adding another `?Trait` to
+      the language, and suggested that having `size_of_val` panic was sufficient (the
+      current implementation does not panic and returns zero instead, which is also
+      deemed undesirable).
+        - It was argued that `?Trait`s are powerful and should be made more
+          ergonomic rather than avoided.
+    - [kennytm][author_kennytm] left a useful comment summarising [which
+      standard library bounds would benefit from relaxation to a `DynSized` 
+      bound](https://github.com/rust-lang/rust/pull/46108#issuecomment-353672604).
+    - Ultimately this was closed [after a language team meeting](https://github.com/rust-lang/rust/pull/46108#issuecomment-360903211)
+      deciding that `?DynSized` was ultimately too complex and couldn't be
+      justified by support for a relatively niche feature like extern types.
+- [rfcs#2255: More implicit bounds (?Sized, ?DynSized, ?Move)][issue_more_implicit_bounds], [kennytm][author_kennytm], Dec 2017
+    - Issue created following [rust#46108][pr_dynsized_rebase] to discuss the
+      complexities surrounding adding new traits which would benefit from relaxed
+      bounds (`?Trait` syntax).
+    - There have been various attempts to introduce new auto traits with
+      implicit bounds, such as `DynSized`, `Move`, `Leak`, etc. Often rejected due to
+      the ergonomic cost of relaxed bounds.
+        - `?Trait` being a negative feature can be confusing to users.
+        - Depending on the specific trait being added as a default bound, downstream crates
+          need to re-evaluate every API to determine if adding `?Trait` makes sense, for
+          each `?Trait` added.
+        - `?Trait` isn't actually fully backwards compatible due to interactions with
+          associated types.
```

Comment:
It is still not clear to me what this refers to -- I think `?Trait` is exactly as backwards compatible as this proposal. Changing the meaning of existing associated type bounds is not backwards-compatible, but that's a different question. All associated types in old-edition code are `Move` and `DynSized` and `Leak` and whatnot. Changing *that* would be backwards-incompatible. Letting new editions write traits with types that are not `Move` is completely fine, and we can use he `?Move` syntax for that.

So I think this point here should be removed, unless someone can substantiate what the actual claim here is.
```suggestion
```

Reply:
Alternatively, we could directly discuss this point:
```suggestion
        - existing associated types have to keep their bounds; adding `?Trait` automatically isn't
          backwards compatible. (However, the same happens with this proposal: relaxing associated
          type bounds below their current meaning is not backwards compatible; that's why the edition
          migration has to turn `?Sized` into `const ValueSized`.)
```

Reply:
Fixed